### ISSUE TITLE
Phase 58.5: DiffusionOrchestrator — End-to-End Generation Pipeline & Flow Matching

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1102]
### Overview
The DiffusionOrchestrator coordinates all Phase 58 components into a unified end-to-end generative pipeline. It manages the complete lifecycle from training data preparation through model training, sampling, and inference optimization. It also integrates advanced paradigms including cascaded diffusion, super-resolution pipelines, flow matching, and rectified flows.

### Architecture
```typescript
interface DiffusionOrchestrator {
  // Pipeline management
  createPipeline(config: PipelineConfig): DiffusionPipeline;
  text2image(prompt: string, config: GenerationConfig): GeneratedOutput;
  image2image(image: Tensor, prompt: string, strength: number): GeneratedOutput;
  inpainting(image: Tensor, mask: Tensor, prompt: string): GeneratedOutput;
  
  // Training coordination
  trainModel(dataset: Dataset, config: TrainingConfig): TrainingResult;
  finetune(baseModel: DiffusionPipeline, dataset: Dataset, method: FinetuneMethod): DiffusionPipeline;
  trainLoRA(baseModel: DiffusionPipeline, dataset: Dataset, rank: number): LoRAWeights;
  
  // Cascaded generation
  cascadedGeneration(prompt: string, stages: CascadeStage[]): Tensor;
  superResolution(lowRes: Tensor, factor: number): Tensor;
  
  // Flow matching
  flowMatchingStep(x0: Tensor, x1: Tensor, t: number): {xt: Tensor, ut: Tensor};
  optimalTransportPairing(batch0: Tensor, batch1: Tensor): PairedBatch;
  rectifiedFlowSample(noise: Tensor, numSteps: number): Tensor;
  
  // Inference optimization
  compilePipeline(pipeline: DiffusionPipeline): OptimizedPipeline;
  quantizeModel(model: Denoiser, bits: 4|8): QuantizedModel;
  enableTensorRT(pipeline: DiffusionPipeline): AcceleratedPipeline;
  tokenMerging(model: Denoiser, ratio: number): MergedModel;
  
  // Monitoring & evaluation
  computeFID(generated: Tensor[], reference: Tensor[]): number;
  computeCLIPScore(images: Tensor[], prompts: string[]): number;
  logMetrics(step: number, metrics: TrainingMetrics): void;
}

interface PipelineConfig {
  type: PipelineType;
  scheduler: NoiseScheduler;
  estimator: ScoreEstimator;
  sampler: SamplingEngine;
  encoder?: LatentDiffusionEncoder;
  resolution: number;             // target resolution
  latentResolution?: number;      // latent spatial dim
  numInferenceSteps: number;      // default sampling steps
  guidanceScale: number;          // default CFG scale
}

enum PipelineType {
  PIXEL_DIFFUSION = 'pixel',           // direct pixel-space
  LATENT_DIFFUSION = 'latent',         // VAE latent space (SD-style)
  CASCADED = 'cascaded',               // multi-stage (Imagen-style)
  FLOW_MATCHING = 'flow_matching',     // continuous normalizing flow
  RECTIFIED_FLOW = 'rectified_flow',   // straight-path flows
  CONSISTENCY = 'consistency'           // single-step generation
}

interface TrainingConfig {
  batchSize: number;
  learningRate: number;
  warmupSteps: number;
  totalSteps: number;
  emaDecay: number;               // 0.9999 typical
  gradientAccumulation: number;
  mixedPrecision: 'fp16' | 'bf16' | 'fp32';
  parameterization: ParameterizationType;
  snrGamma?: number;              // Min-SNR weighting
  noiseOffset?: number;           // noise offset for dark/bright images
  inputPerturbation?: number;     // Imagen-style input noise
}

enum FinetuneMethod {
  FULL = 'full',
  LORA = 'lora',
  DREAMBOOTH = 'dreambooth',
  TEXTUAL_INVERSION = 'textual_inversion',
  CONTROLNET = 'controlnet',
  IP_ADAPTER = 'ip_adapter'
}

interface CascadeStage {
  model: DiffusionPipeline;
  inputResolution: number;
  outputResolution: number;
  noiseAugmentation?: number;     // conditioning augmentation level
}

interface GeneratedOutput {
  images: Tensor;
  latents?: Tensor;
  metadata: {
    prompt: string;
    seed: number;
    numSteps: number;
    guidanceScale: number;
    generationTime: number;
    modelId: string;
  };
}

interface TrainingMetrics {
  loss: number;
  gradNorm: number;
  learningRate: number;
  emaDecay: number;
  samplesPerSec: number;
  fid?: number;
  clipScore?: number;
}
```

#### `text2image()`
Complete text-to-image pipeline: encode text → sample noise → iterative denoising with CFG → decode latent → post-process.

#### `flowMatchingStep()`
Flow matching training: x_t = (1-t)·x_0 + t·x_1, predict velocity u_t = x_1 - x_0. Simpler than score matching, enables straight ODE paths.

#### `cascadedGeneration()`
Multi-stage generation: base model generates at 64×64, stage-2 upscales to 256×256, stage-3 to 1024×1024, with noise augmentation between stages.

#### `trainLoRA()`
Low-rank adaptation: inject trainable rank-r decomposition matrices into attention layers, enabling efficient fine-tuning with <1% parameters.

#### `computeFID()`
Fréchet Inception Distance evaluation: measures distribution similarity between generated and real images using Inception-v3 features.

### Implementation Details
- Modular pipeline assembly — swap components without code changes
- Automatic mixed precision with loss scaling
- Distributed training support (data parallel + model parallel)
- Dynamic batch size based on available VRAM
- Inference: ONNX export, TensorRT compilation, torch.compile
- Token merging (ToMe) for 2× inference speedup
- Watermarking for generated content attribution
- Safety checker integration for content filtering

### References
- Rombach et al. 2022 — Latent Diffusion / Stable Diffusion pipeline
- Saharia et al. 2022 — Imagen cascaded pipeline
- Lipman et al. 2023 — Flow Matching for generative modeling
- Liu et al. 2023 — Rectified Flow / InstaFlow
- Song et al. 2023 — Consistency Models for single-step generation
- Hu et al. 2022 — LoRA for efficient fine-tuning
- Ruiz et al. 2023 — DreamBooth subject-driven generation
- Bolya & Hoffman 2023 — Token Merging (ToMe)

### Dependencies
- Phase 58.1 NoiseScheduler — noise schedule management
- Phase 58.2 ScoreEstimator — denoiser network
- Phase 58.3 SamplingEngine — sampling algorithms
- Phase 58.4 LatentDiffusionEncoder — VAE and conditioning
- Phase 45 (Model Compression) — quantization and distillation

### Testing Requirements
- End-to-end text2image generation produces valid images
- Cascaded pipeline improves resolution without artifacts
- Flow matching training converges with fewer steps than DDPM
- LoRA fine-tuning preserves base model quality
- FID evaluation correlates with visual quality assessment
- Inference optimization achieves ≥2× speedup with <5% quality loss

**Comment by web3guru888** [comment]
## Phase 59 — Parameter-Efficient Fine-Tuning & Adaptation (PEFT) — LAUNCHED 🚀

Phase 58 (Diffusion Models) successor now live. Phase 59 introduces a comprehensive PEFT framework:

- **59.1 LoRAEngine** (#1114) — Low-rank adaptation with LoRA/DoRA/AdaLoRA variants
- **59.2 AdapterManager** (#1115) — Bottleneck & parallel adapters, AdapterFusion, MAD-X
- **59.3 PromptTuner** (#1116) — Soft prompt tuning, prefix tuning, P-Tuning v2
- **59.4 QuantizedFineTuner** (#1117) — QLoRA 4-bit, GPTQ/AWQ backends
- **59.5 PEFTOrchestrator** (#1118) — Unified pipeline with method selection & serving

Planning discussion: #1113
